c10/core/Device.h

```diff
@@ -16,7 +17,10 @@ namespace c10 {
 /// A DeviceIndex is not independently meaningful without knowing
 /// the DeviceType it is associated; try to use Device rather than
 /// DeviceIndex directly.
-using DeviceIndex = int8_t;
+using DeviceIndex = int16_t;
+
+/// The last, still valid device index
+constexpr int64_t MAX_DEVICE_INDEX = std::numeric_limits<DeviceIndex>::max();
```

Comment:
Why not use 512 here and set C10_COMPILE_TIME_MAX_GPUS to this value?

Reply:
I was thinking about this too, but decided not to couple the amount of GPUs to the max possible index in case we want to support things like `meta` and `lazy` tensors with >512 indices. I don't know if that's a realistic use-case though, I can also just reuse `C10_COMPILE_TIME_MAX_GPUS`.

Reply:
I think the same argument for limiting max GPUs for now applies here. I'm sure someone will soon use this constant to store a static array :p 
It's an easy one to flip so I would prefer all the devices to be consistent.

Reply:
I now introduced a new constant `C10_MAX_NUM_DEVICES` and set `C10_COMPILE_TIME_MAX_GPUS` based on it. Maybe we should just remove `C10_COMPILE_TIME_MAX_GPUS` completely from `CUDAMacros.h` and replace everything with `C10_MAX_NUM_DEVICES`.